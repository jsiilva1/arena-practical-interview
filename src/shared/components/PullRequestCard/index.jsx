/**
 * External Dependencies
 */
import React from 'react';
import Skeleton from 'react-loading-skeleton';

/**
 * Internal Dependencies
 */
import Icon from '../Icon';
import { formatDate } from '../../utils/helpers';
import {
  PullRequestItem,
  PullRequestItemTitle,
  PullRequestItemMetadata,
  PullRequestUser,
} from './styles/index';

/**
 * PullRequestCard Component
 *
 * @description render a card of pull request
 *
 * @param {Object} props props of component
 *
 * @return {*} component.
 */
const PullRequestCard = (props) => {
  // Extract all props
  const {
    id,
    title,
    number,
    user,
    link,
    date,
    state,
    isLoading,
  } = props;

  // Segment any properties to avoid errors of null cases
  const body = props.body ? props.body : 'no description';

  let username;
  let avatar_url;

  if (user) {
    username = user.login;
    avatar_url = user.avatar_url;
  }

  return (
    <>
      <a href={link} target="_blank" rel="noopener noreferrer" key={id}>
        <PullRequestItem>
          <PullRequestItemTitle>
            <span>
              {isLoading ? (
                <span style={{ marginRight: '10px' }}>
                  <Skeleton width={30} height={15} />
                </span>
              ) : (
                <Icon
                  className="icon"
                  name={['fas', 'code-branch']}
                  vendor="fa"
                  color={state === 'open' ? '#40A745' : '#CB2431'}
                  title={state}
                />
              )}

              {isLoading ? (
                <Skeleton width={640} height={15} />
              ) : <span>#{number} {title}</span>}
            </span>
          </PullRequestItemTitle>

          <PullRequestItemMetadata>
            {isLoading ? (
              <Skeleton count={5} />
            ) : body}
          </PullRequestItemMetadata>

          <PullRequestUser>
            <span className="user-photo">
              {isLoading ? (
                <Skeleton circle width={20} height={20} />
              ) : <img src={avatar_url} alt="User" />}
            </span>

            <span className="metadata-text-info">
              {isLoading ? (
                <Skeleton width={300} height={15} />
              ) : (
                <>
                  <b className="username">{username}</b>
                  <span className="date"> â€” {formatDate(date)}</span>
                </>
              )}
            </span>
          </PullRequestUser>
        </PullRequestItem>
      </a>
    </>
  );
};

export default PullRequestCard;
