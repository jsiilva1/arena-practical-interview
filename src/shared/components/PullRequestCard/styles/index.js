import styled from 'styled-components';

export const PullRequestItem = styled.div`
  border: 1.5px solid #F0F0F0;
  background-color: #FFF;
  margin: 5px 0;
  padding: 7px 15px;
  font-size: 1rem;
  display: flex;
  flex-direction: column;
`;

export const PullRequestItemTitle = styled.span`
  display: flex;
  flex-direction: row;
  font-size: .9rem;
  font-weight: 800;
  margin-top: 5px;

  & .icon {
    margin-right: 8px;
  }

  & span {
    margin: 0;
  }
`;

export const PullRequestItemMetadata = styled.span`
  margin: 7px 0 0 18px;
  font-size: .8rem;
  color: #586069;
  max-width: 100vh;
  display: block;
  text-align: justify;
  text-overflow: ellipsis;
  word-wrap: break-word;
  overflow: hidden;
  max-height: 7.1em;
  line-height: 1.2em;
  -webkit-line-clamp: 4;
  -webkit-box-orient: vertical;
`;

export const PullRequestUser = styled.span`
  display: flex;
  flex-direction: row;
  margin-top: 10px;
  margin-left: 18px;
  font-size: 0.8rem;

  & .user-photo {
    width: 20px;
    height: 20px;

    & img {
      border-radius: 50%;
      max-width: 100%;
      max-height: 100%;
    }
  }

  & .metadata-text-info {
    margin: 1px 5px;
    color: #666;
  }

  & .username {
    font-size: 1rem;
    margin: 0;
  }

  & .date {
    position: relative;
    top: -1px;
    left: 1px;
    font-size: .75rem;
  }  
`;
