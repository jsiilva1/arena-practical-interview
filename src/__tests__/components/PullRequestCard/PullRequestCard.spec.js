import React from 'react';
import { shallow } from 'enzyme';

import PullRequestCard from '../../../shared/components/PullRequestCard';

describe('PullRequestCard Component', () => {
  it('should render a Pull Request Card correctly', () => {
    const wrapper = shallow(
      <PullRequestCard
        id={1}
        title="pull request title"
        number={5422}
        user={{ login: 'jsiilva1', avatar_url: null }}
        date={"Mon Dec 23 2019 13:16:40 GMT-0300 (Horário Padrão de Brasília)"}
        link={"null"}
        body={"content of pull request"}
        state={'open'}
      />,
    );
    
    expect(wrapper).toMatchSnapshot();
  });
});
