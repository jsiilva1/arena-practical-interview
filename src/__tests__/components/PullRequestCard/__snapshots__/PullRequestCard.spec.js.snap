// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PullRequestCard Component should render a Pull Request Card correctly 1`] = `
<Fragment>
  <a
    href="null"
    key="1"
    rel="noopener noreferrer"
    target="_blank"
  >
    <styled.div>
      <styled.span>
        <span>
          <Icon
            className="icon"
            color="#40A745"
            name={
              Array [
                "fas",
                "code-branch",
              ]
            }
            title="open"
            vendor="fa"
          />
          <span>
            #
            5422
             
            pull request title
          </span>
        </span>
      </styled.span>
      <styled.span>
        content of pull request
      </styled.span>
      <styled.span>
        <span
          className="user-photo"
        >
          <img
            alt="User"
            src={null}
          />
        </span>
        <span
          className="metadata-text-info"
        >
          <b
            className="username"
          >
            jsiilva1
          </b>
          <span
            className="date"
          >
             — 
            1:16 PM · Dec 23, 2019
          </span>
        </span>
      </styled.span>
    </styled.div>
  </a>
</Fragment>
`;
